### ISSUE TITLE
Not working on 1.21.0

### ISSUE THREAD

**Issue by ghost** [opl-/beatsaber-http-status#86]
## Current behavior
Plugin does not initialize or function on BS 1.21.0 (probably you're already working on an update, but in case not I wanted to enter an issue for it).  Errors shown in log:
```
[WARNING @ 22:23:40 | IPA/Loader] Mod Beat Saber HTTP Status developed for game version 1.19.0, so it may not work properly.
[ERROR @ 22:23:40 | IPA/Loader] Error creating executor for Beat Saber HTTP Status
[ERROR @ 22:23:40 | IPA/Loader] System.TypeLoadException: Could not resolve type with token 01000034 (from typeref, class/assembly ICutScoreBufferDidFinishEvent, Main, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null)
[ERROR @ 22:23:40 | IPA/Loader]   at (wrapper managed-to-native) System.Reflection.Assembly.InternalGetType(System.Reflection.Assembly,System.Reflection.Module,string,bool,bool)
[ERROR @ 22:23:40 | IPA/Loader]   at System.Reflection.MonoAssembly.GetType (System.String name, System.Boolean throwOnError, System.Boolean ignoreCase) [0x00022] in <eae584ce26bc40229c1b1aa476bfa589>:0 
[ERROR @ 22:23:40 | IPA/Loader]   at System.Reflection.Assembly.GetType (System.String name) [0x00000] in <eae584ce26bc40229c1b1aa476bfa589>:0 
[ERROR @ 22:23:40 | IPA/Loader]   at IPA.Loader.PluginExecutor.PrepareDelegates () [0x00026] in <85e5e5773585418d89fcc2712ee48bcc>:0 
[ERROR @ 22:23:40 | IPA/Loader]   at IPA.Loader.PluginExecutor..ctor (IPA.Loader.PluginMetadata meta, IPA.Loader.PluginExecutor+Special specialType) [0x00087] in <85e5e5773585418d89fcc2712ee48bcc>:0 
[ERROR @ 22:23:40 | IPA/Loader]   at IPA.Loader.PluginLoader.InitPlugin (IPA.Loader.PluginMetadata meta, System.Collections.Generic.IEnumerable`1[T] alreadyLoaded) [0x00118] in <85e5e5773585418d89fcc2712ee48bcc>:0 
```

## Expected behavior
Plugin initializes and functions as in BS 1.19.0 and below

## Reproduction steps
1. Install plugin and prerequisites
2. Launch game
3. Notice errors in log and no functionality from the plugin

## Environment

BeatSaber version:  1.21.0
HTTP Status version:  1.20.0
Platform:  SteamVR

**Comment by Raiden38** [comment]
Same issue here, hope it will be fixed soon, it looked very good in a stream! :)

**Comment by opl-** [comment]
I already made an attempt at updating for this version (see the [update/1.20.0](https://github.com/opl-/beatsaber-http-status/tree/update/1.20.0) branch) and in the process found some rather unfortunate changes

The short version is that this update has changed the scoring system significantly enough that it'd require significant changes to this mod to keep the data it sends make sense. As an example, using the current version of the branch above the score ends up showing the score with a single cut delay: if there's no notes to cut for half a minute, overlays will continue showing the now out-of-date score for half a minute, because the score for the last cut was never sent out.

More specifically, scoring now happens in bulk at the end of the frame for each note cut during that frame. There's no events sent inside of the loop which iterates over cuts to score during that frame either, which further complicates figuring out the score for each of the cuts.

This mod has been long due for a rewrite but I have no motivation to do so, especially with how life's going right now.

For the time being you can try denpadokei's fork (they don't seem to know English, so there might be some setup differences, documentation in Japanese, or missing documentation): https://github.com/denpadokei/HttpSiraStatus/releases/tag/7.0.0

**Comment by craftxbox** [comment]
Is there any chance you could temporarily/indefinately 'neuter' the score reporting section? Even without scoring this mod has been really useful for track data overview, The scoring info isn't as important IMO.

**Comment by TakingFire** [comment]
Likewise, HueSaber only relies on lighting events to function. Assuming your attempt is functional in that regard, could you possibly provide the compiled plugin?
I took a look at the forked version, and while it does _work_ in the latest version, it seems to serve junk data as far as beatmapEvent goes. (As well as some missing `type` that contain the performance object?) It unfortunately is not useable at least at first glance.

**Comment by opl-** [comment]
The events code is completely commented out in that branch because the difficulty `.dat` files are now on version 3 (community docs [here](https://bsmg.wiki/mapping/map-format.html#difficulty-file-v3)), which completely changes how events work.

I believe I came to the conclusion that a lot of it is no longer backwards compatible with the old events protocol, to the point where I wasn't sure if I could even translate back to the old format at all.

Internally, the game automatically converts the old 2.x format to the new format, meaning even sending events for older maps that are still in the old 2.x format is non-trivial.

I genuinely think that at this point it might be easier to just rewrite the whole mod from scratch built on top of Zenject and MonoMod with a redesigned protocol, than it would be to make the old protocol work with all these changes. That said, this would of course make all the overlays stop working (at least without putting even more effort into building a translation layer for backwards compatibility) and it still wouldn't be an evening project.

**Comment by Protected** [comment]
I'll fix my overlay if anyone rewrites and provides a reasonably equivalent interface. Let me know here on in the note history repository.

**Comment by craftxbox** [comment]
I looked into [HttpSiraStatus](https://github.com/denpadokei/HttpSiraStatus) and it appears to function as a drop-in replacement, already working on 1.22.1. No functional differences were encountered, however, the [overlay I use personally](https://crxb.cc/bsoverlay/) (a copy of a project I can't quite remember the repository for at the moment) only uses a small subset of information, so this is by no means a complete assessment of its functionality.

**Comment by Protected** [comment]
I'll try it out next time I play!

**Comment by Protected** [comment]
HttpSiraStatus at first glance doesn't seem to have incorporated the improvements we made in late 2021. These are important, as the old protocol delivers wrong and incomplete information for note cuts.

**Comment by Morgolf** [comment]
Any updates on this? I have an overlay that doesn't work with Sira version too well (its an overlay that shows hit score visualizer and slice visualizer as a browser source and according to a comment above, that data is incorrect) :(

**Comment by Protected** [comment]
I've been able to test yesterday with the current version of HttpSiraStatus, 7.1.5 (I think I downloaded the release bundle, not 100% sure).

Currently, it seems to be working reasonably well for note events! Worst case, the setup might be failing to keep up when the level is very dense (I've noticed it here and there); I could always use more opinions on that though since it's hard to tell.

For example, [in this map the note history seems to be in sync from start to end](https://www.twitch.tv/videos/1523366102?t=00h20m55s).

But [half-way through this map something feels off](https://www.twitch.tv/videos/1523366102?t=01h04m22s), especially with the left hand, like it's completely off time or some of the notes haven't been tracked (both hands use the same code in Note History and the trackers are independent in OBS so that makes no sense).

As for the score, it seems correct? Matches the end of level score and the ranks seem correct. Way better than the currently very broken counters+ at least.

I did not play any levels with the new note types yesterday, so don't bother looking for them in the video. I shudder to think of how horribly they're going to break the whole thing though!

**Comment by Morgolf** [comment]
Yeah the score is correct, I have the same issue with the overlay not keeping up which was not happening with the original non sira status back on 18.3 version of the game for example. Heres an example of my overlay breaking with Sira, notice the bottom right of the screen, I put the overlay over the camera to create this slice visualize effect + hit scores https://www.twitch.tv/videos/1524287990

**Comment by Protected** [comment]
I see it, yeah. So it really can't keep up, which is a bummer. Should we open a ticket over at the httpsirastatus repo?

> according to a comment above, that data is incorrect

According to the research we did last year, that way of drawing slices is too naive to be useful; IIRC it just calculates inclination from the entry point and draws a line through the center of the front face of the note, or something like that. It's an incorrect representation of the cut.

In fact the vast majority of cuts don't cut through the center of the note. Distance to the center is the most important factor in determining cut score (or was, at least), so you should only see perfect or near perfect cuts drawn through the center (plus a 15 point bonus for wide arcs, new note types notwithstanding). Inclination doesn't matter at all other than to determine whether you cut in a direction sufficiently consistent with the arrow on the cube (it's a binary thing). It matters more if strict angles is on, but it's still a binary thing.

Also important, unless you enable pro mode you can cut entirely outside the note, as there's a tolerance, especially on the sides of the note (left and right). These cuts are valid but shouldn't have slashes drawn through them at all! [Here's opl's mockup of the real cut volume as of 1.18](https://cdn.discordapp.com/attachments/897074382692835350/899122394680147999/unknown.png)

In my overlay you can see an accurate representation of the 3D projection of the cut on the note by using [showCutBacks=1&showCutSides=1](https://github.com/Protected/beatsaber-notehistory/wiki/Settings#note-rendering-settings) (not enabled in the video above). Feel free to copy the javascript if you need it. opl did nearly all of the math!

**Comment by Morgolf** [comment]
If you have the knowledge to do a proper helpful ticket, would be great if you would do it on Sira status, maybe that would help. My overlay was made by another person so I am not sure how exactly it is structured, I remember the dev mention the issue with representing the cuts away from the middle, he did some calculations based off of the cut score if I remember correctly? I am not entirely sure but I remember the cuts working correctly, we compared them to the in game slice visualizer mod and it was accurate, not sure if it works correctly with the way Sira works now or not. This is the project for the overlay I am using, I just run it locally and use localhost for browser source https://github.com/EmilMalanczak/beat-saber-overlay
